cpp/src/arrow/compute/exec/asof_join_node_test.cc

```diff
@@ -73,8 +74,9 @@ Result<BatchesWithSchema> MakeBatchesFromNumString(
     const std::vector<std::string_view>& json_strings, int multiplicity = 1) {
   FieldVector num_fields;
   for (auto field : schema->fields()) {
-    num_fields.push_back(
-        is_base_binary_like(field->type()->id()) ? field->WithType(int64()) : field);
+    auto id = field->type()->id();
+    bool adjust = id == Type::BOOL || is_base_binary_like(id);
+    num_fields.push_back(adjust ? field->WithType(int64()) : field);
```

Comment:
Not sure I understand this - what does `field->WithType(int64())` do here?

Reply:
This instructs the JSON parsing to interpret the field as an `int64` value even though the schema of the test case defines the field as a binary type. This is because all values in the test data are numeric, so in order to get a binary type (such as a string "102" instead of the number 102) we first parse it as a number, and later on [convert the number to the binary type](https://github.com/westonpace/arrow/pull/24/files#diff-6f11dfb4e3adce81ffca316bab039452b80c41a2453fdf6fe9df1b3085d8032dR95-R97).